src/fides/api/util/consent_util.py

```diff
@@ -85,15 +95,39 @@ def should_opt_in_to_service(
             [],  # Don't need to return the filtered preferences, this is just relevant for the new workflow
         )
 
-    # NEW WORKFLOW
-    relevant_preferences = filter_privacy_preferences_for_propagation(
-        system,
-        privacy_request.privacy_preferences,  # type: ignore[attr-defined]
+    # NEW WORKFLOWS: 1.notice-based and 2. global SaaS Consent
+    relevant_preferences: List[PrivacyPreferenceHistory] = (
+        filter_privacy_preferences_for_propagation(
+            system,
+            privacy_request.privacy_preferences,  # type: ignore[attr-defined]
+        )
     )
     if not relevant_preferences:
         return None, []  # We should do nothing here
 
-    # Collapse relevant preferences into whether we should opt-in or opt-out
+    filtered_preferences: List[PrivacyPreferenceHistory] = []
+
+    # 1. NOTICE-BASED WORKFLOW
+    if is_notice_based:
+        notice_id_to_preference_map: Dict[str, UserConsentPreference] = {}
+        # retrieve notices from the DB if they are associated with a relevant preference
+        notices_with_preference: Query = session.query(PrivacyNotice).filter(
```

Comment:
This chunk of work attempts to build our notice map and preferences in the most efficient way possible, in which we only need to make one DB query instead of multiple.